pkg/agent/openflow/pod_connectivity.go

```diff
@@ -155,16 +161,92 @@ func (f *featurePodConnectivity) initFlows() []binding.Flow {
 		flows = append(flows, getCachedFlows(cachedFlows)...)
 	}
 
 	return flows
 }
+
+// trafficControlMarkFlows generates the flows to mark the packets that need to be redirected or mirrored.
+func (f *featurePodConnectivity) trafficControlMarkFlows(sourceOFPorts []uint32, targetOFPort uint32, direction v1alpha2.Direction, action v1alpha2.TrafficControlAction) []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	var actionRegMark *binding.RegMark
+	if action == v1alpha2.ActionRedirect {
+		actionRegMark = TrafficControlRedirectRegMark
+	} else if action == v1alpha2.ActionMirror {
+		actionRegMark = TrafficControlMirrorRegMark
+	}
+	var flows []binding.Flow
+	for _, port := range sourceOFPorts {
+		if direction == v1alpha2.DirectionIngress || direction == v1alpha2.DirectionBoth {
+			// This generates the flow to mark the packets destined for a provided port.
+			flows = append(flows, TrafficControlTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchRegFieldWithValue(TargetOFPortField, port).
+				Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
+				Action().LoadRegMark(actionRegMark).
+				Action().NextTable().
+				Done())
+		}
+		// This generates the flow to mark the packets sourced from a provided port.
+		if direction == v1alpha2.DirectionEgress || direction == v1alpha2.DirectionBoth {
+			flows = append(flows, TrafficControlTable.ofTable.BuildFlow(priorityNormal).
+				Cookie(cookieID).
+				MatchInPort(port).
+				Action().LoadToRegField(TrafficControlTargetOFPortField, targetOFPort).
+				Action().LoadRegMark(actionRegMark).
+				Action().NextTable().
+				Done())
+		}
+	}
+	return flows
+}
+
+// trafficControlReturnClassifierFlow generates the flow to mark the packets from traffic control return port.
+func (f *featurePodConnectivity) trafficControlReturnClassifierFlow(returnOFPort uint32) binding.Flow {
+	return ClassifierTable.ofTable.BuildFlow(priorityNormal).
+		Cookie(f.cookieAllocator.Request(f.category).Raw()).
+		MatchInPort(returnOFPort).
+		Action().LoadRegMark(FromTCReturnRegMark).
+		Action().GotoStage(stageRouting).
+		Done()
+}
+
+// trafficControlCommonFlows generates the common flows for traffic control.
+func (f *featurePodConnectivity) trafficControlCommonFlows() []binding.Flow {
+	cookieID := f.cookieAllocator.Request(f.category).Raw()
+	return []binding.Flow{
+		// This generates the flow to output packets to the original target port as well as mirror the packets to the target
+		// traffic control port.
+		L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh+1).
+			Cookie(cookieID).
+			MatchRegMark(OFPortFoundRegMark, TrafficControlMirrorRegMark).
+			Action().OutputToRegField(TargetOFPortField).
+			Action().OutputToRegField(TrafficControlTargetOFPortField).
+			Done(),
+		// This generates the flow to output the packets to be redirected to the target traffic control port.
+		L2ForwardingOutTable.ofTable.BuildFlow(priorityHigh+1).
```

Comment:
Question: A packet is either mirrored or redirected, but not both mirrored and redirected?

Reply:
A packet should be either mirrored or redirected.